### PR TITLE
[Deepspeed Wav2vec2] integration

### LINKED ISSUE #11446: [wav2vec] deepspeed eval bug in the case of >1 gpus
## Environment info

- `transformers` version: 4.5.1
- Platform: Linux-4.15.0-140-generic-x86_64-with-debian-buster-sid
- Python version: 3.7.9
- PyTorch version (GPU?): 1.8.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: <2,4>
- Using distributed or parallel set-up in script?: <distributed>

### Who can help
@stas00 
@patrickvonplaten 
@patil-suraj 

## Information

I'm working on wav2vec2.0 using the following official script of huggingface.
https://github.com/huggingface/transformers/blob/master/examples/research_projects/wav2vec2/run_common_voice.py

I am trying to finetune huggingface model with multiple gpus using deepspeed.

```
deepspeed --num_gpus=1 run_common_voice.py --deepspeed ds_config.json --do_train --do_eval
```
works, but

```
deepspeed --num_gpus=2 run_common_voice.py --deepspeed ds_config.json --do_train --do_eval
```

stops working and freezes at the end of eval. 
The progress bar is 100% done but the eval result is not returned and it freezes.

## To reproduce

This is how to reproduce!
https://colab.research.google.com/drive/1VRCGcnhBlrMFYQ5aaNebucZuja-WB2I2?usp=sharing

1. Install deepspeed
2. Add `with autocast():` after line 481 in run_common_voice.py
3. Set param: `--deepspeed ds_config.json --do_train --do_eval`
4. Run run_common_voice.py using deepspeed with 1> gpus

ds_config has the following parameters.
```ds_config.json
{
  "fp16": {
    "enabled": "true",
    "loss_scale": 0,
    "loss_scale_window": 1000,
    "hysteresis": 2,
    "min_loss_scale": 1,
    "opt_level": "O3"
  },
  "steps_per_print": 100,
  "wall_clock_breakdown": "false"
}
```

## Expected behavior

The finetuning eval should be executed without freezing.

### PR DESCRIPTION
**Pull request by stas00**
Addressing the need in https://github.com/huggingface/transformers/issues/11446, this PR is working on making wav2vec2 work under deepspeed.

This PR:
* changes Trainer to automatically convert inputs to the correct dtype if it's not int64 - we didn't need this for nlp models because embeddings took care of this - this is not the case with wav2vec2 type of models where inputs are float32 by default. (for deepspeed only at the moment - potentially need to do the same for `fp16_full_eval`)
* multiple fixes to the `wav2vec2` model, because it does very non-standard things, like model `weight_norm` which is implemented in a very odd way and deepspeed's automatic ease-of-use fails to do that and requires multiple manual adjustments for it to do the right thing.  `weight_norm`  creates a param, then drops it replacing it with 2 other params and re-creates them on every forward. (in pre-hook).
* moves `require_deepspeed` to `testing_utils.py` as we have multiple test files using it
* adds `dtype` accessor to DS conf object
* adds 8 new tests, checking each setup

Testing with `run_asr.py`:

### ZeRO-2 

Everything works:

* [x] fp16 distributed zero2
* [x] fp16 non distributed zero2
* [x] fp32 distributed zero2
* [x] fp32 non distributed zero2

important - must use for distributed use:
```
    "zero_optimization": {
        "find_unused_parameters": true,
```
So you can use the `--deepspeed examples/research_projects/wav2vec2/ds_config_wav2vec2_zero2.json` which already has the adjustment.

### ZeRO-3 

You can use the `--deepspeed examples/research_projects/wav2vec2/ds_config_wav2vec2_zero3.json`

This works:

* [x] fp32 non distributed zero3
* [x] fp32 distributed zero3 
* [x] fp16 non distributed zero3
* [x] fp16 distributed zero3

### Possible PR spin-offs

it looks like plain pytorch dist doesn't work either https://github.com/huggingface/transformers/issues/11452
so this PR can be adapted to detect `dist` and do the same as what deepspeed branch does. probably a separate PR is the best.
(LayerSkip that is)

---------------------

To run tests:

Install this deepspeed master https://github.com/microsoft/DeepSpeed:
```
pip install deepspeed
```

 and then:

```
HF_DATASETS_IN_MEMORY_MAX_SIZE=0 RUN_SLOW=1 pyt examples/research_projects/wav2vec2/test_wav2vec2_deepspeed.py

```

---------------------
Example of usage: assuming you in a top dir of the git clone of this branch

### `run_asr.py` and tiny model and tiny dataset

This is the foundation for the new tests:

```
HF_DATASETS_IN_MEMORY_MAX_SIZE=0 PYTHONPATH=src deepspeed --num_gpus 2 \
examples/research_projects/wav2vec2/run_asr.py \
--output_dir=output_dir --num_train_epochs=2 --per_device_train_batch_size=2 \
--per_device_eval_batch_size=2 --evaluation_strategy=steps --save_steps=500 --eval_steps=100 \
--logging_steps=5 --learning_rate=5e-4 --warmup_steps=3000 \
--model_name_or_path=patrickvonplaten/wav2vec2_tiny_random_robust \
--dataset_name=patrickvonplaten/librispeech_asr_dummy --dataset_config_name=clean \
--train_split_name=validation --validation_split_name=validation --orthography=timit \
--preprocessing_num_workers=1 --group_by_length --freeze_feature_extractor --verbose_logging \
--deepspeed examples/research_projects/wav2vec2/ds_config_wav2vec2_zero2.json
```

### run_common_voice.py

very hard to test with as it takes some 5-10mins to just get ready to run.

**edit**: switch to `datasets` master branch and add `HF_DATASETS_IN_MEMORY_MAX_SIZE=0` to the command line - it will be cached now.

`run_common_voice.py` now runs under `--fp16` but gives `loss=nan`, probably the same issue as bf16-pretrained models? I tested - it has the same issue under AMP and no deepspeed. So it's a different problem to solve.

fp32 works just fine loss-wise, you can try:

```
HF_DATASETS_IN_MEMORY_MAX_SIZE=0  PYTHONPATH="src" deepspeed --num_gpus=1 \
examples/research_projects/wav2vec2/run_common_voice.py \
--model_name_or_path="facebook/wav2vec2-large-xlsr-53" --dataset_config_name="tr" \
--output_dir=./wav2vec2-large-xlsr-turkish-demo --overwrite_output_dir --num_train_epochs="5" \
--per_device_train_batch_size="16" --learning_rate="3e-4" --warmup_steps="500" \
--evaluation_strategy="steps" --save_steps="5" --eval_steps="5" --logging_steps="5" \
--save_total_limit="3" --freeze_feature_extractor --feat_proj_dropout="0.0" --layerdrop="0.1" \
--gradient_checkpointing --group_by_length --do_train --do_eval --deepspeed \
examples/research_projects/wav2vec2/ds_config_wav2vec2_zero2.json 
```

Thanks to @patrickvonplaten for making small wav2vec2 models which helped a ton to debug faster and they were needed for the tests.

## Requirements to merge this PR

- [x] https://github.com/microsoft/DeepSpeed/pull/1135
- [x] deepspeed version requirement bumped to 0.4.0
- [x] deepspeed 0.4.0 released

Fixes: https://github.com/huggingface/transformers/issues/11446

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Also link: #11452 here

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1701,7 +1701,14 @@ def _prepare_inputs(self, inputs: Dict[str, Union[torch.Tensor, Any]]) -> Dict[s
...
             if isinstance(v, torch.Tensor):
-                inputs[k] = v.to(self.args.device)
+                kwargs = dict(device=self.args.device)
+                if self.deepspeed and inputs[k].dtype != torch.int64:
+                    # NLP models inputs are int64 and those get adjusted to the right dtype of the
+                    # embedding. Other models such as wav2vec2's inputs are already float and thus
+                    # may need special handling to match the dtypes of the model
+                    kwargs.update(dict(dtype=self.args.hf_deepspeed_config.dtype()))
+
+                inputs[k] = v.to(**kwargs)
```
Comment (stas00) [inline src/transformers/trainer.py:1711]: @sgugger, this one would be of main interest to you.

Comment (sgugger) [reply src/transformers/trainer.py:1711]: I am not against adding this as a principle, but isn't this a fault of DeepSpeed that should be fixed there? PyTorch AMP does not need us to do this manually.

**Comment by stas00** [comment]
Also while working on this I came up with a fused version of  `Conv1d` + `WeightNorm` which is simpler to understand and work with as compared to the original which uses a pre-forward hook, but I'm not sure how to test its correctness - I only tested it in my head.

I will paste it here for now should we want to try it in the future.

```
import torch.nn as nn
from torch.nn.parameter import Parameter
from torch import _weight_norm, norm_except_dim
class Conv1dWithWeightNorm(nn.Conv1d):
    def __init__(self, *args, **kwargs):
        super(Conv1dWithWeightNorm, self).__init__(*args, **kwargs)
        self.dim = 2
        import deepspeed
        with deepspeed.zero.GatheredParameters(self.weight):
            weight = self.weight
        self.weight_g = Parameter(norm_except_dim(weight, 2, self.dim).data)
        self.weight_v = Parameter(weight.data)
        del self._parameters["weight"]
        self.weight = _weight_norm(self.weight_v, self.weight_g, self.dim)
        print(self.weight)

    def compute_weight(self):
        self.weight_g = Parameter(norm_except_dim(self.weight, 2, self.dim).data)
        self.weight_v = Parameter(self.weight.data)
        return _weight_norm(self.weight_v, self.weight_g, self.dim)

    def forward(self, input):
        self.weight = self.compute_weight()
        return self._conv_forward(input, self.weight, self.bias)
```

File: src/transformers/trainer.py (hunk 1 above)
Comment (stas00) [reply src/transformers/trainer.py:1711]: I am not sure my comment communicated the problem. 

- Under NLP models you get inputs of `int` and they get converted to the correct variation of `float` via the embedding amp - so we get fp32 if the model is normal, or fp16 if it's `model.half()` 
- Here the inputs are `float32` - how could deepspeed possibly guess what users inputs are supposed to be? It's the Trainer that needs to set `inputs.to(dtype)`. e.g. how could it know what to do with inputs of `int64`
- AMP doesn't change the dtype of user inputs, since it never works with fp16 directly unless the user set the inputs to it and only adjusts things dynamically to fp16 during some of the computations. 

If you have an idea of how this deriving can be automated, I'm all ears. I'm just not sure how Deepspeed could possibly know what user has in mind.

Do you perhaps propose that if the inputs are fp32 then Deepspeed can automatically convert them to fp16 if it's doing its amp version? But if inputs are anything else than fp32 (e.g. int64`) it should leave them alone?

Comment (stas00) [reply src/transformers/trainer.py:1711]: we are also getting ready for new dtypes with the bleed edge pytorch - fp16, tf32, etc. so we should think globally here.

Comment (sgugger) [reply src/transformers/trainer.py:1711]: > - AMP doesn't change the dtype of user inputs, since it never works with fp16 directly unless the user set the inputs to it and only adjusts things dynamically to fp16 during some of the computations.

That is incorrect. Inside the autocast context manager, the conversion is done automatically, so the user does not have to think about the dtype of the inputs. In the [official AMP tutorial](https://pytorch.org/docs/stable/notes/amp_examples.html#typical-mixed-precision-training) there is no inputs dtype conversion and having int inputs is a specificity of NLP, not the general rule. All CV models have float inputs and I think that is still the majority of Deep Learning models.

DeepSpeed could also properly deal with dtype of the inputs since they wrap the model in their own container. I'm just noticing a point where they could make their API better and more aligned with PyTorch, I have no problem with merging this like I said before.

Comment (stas00) [reply src/transformers/trainer.py:1711]: Yes, yes, I'm not arguing, just trying to understand how they could do it in a safe way. We may merge this to move wav2vec2 along and meanwhile I'm talking to Deepspeed about automating this and once it's there to drop my workaround.

> That is incorrect. Inside the autocast context manager, the conversion is done automatically, so the user does not have to think about the dtype of the inputs. In the official AMP tutorial there is no inputs dtype conversion and having int inputs is a specificity of NLP, not the general rule. All CV models have float inputs and I think that is still the majority of Deep Learning models.

I still don't think we are on the same page for AMP. 

I stepped through https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html in debugger and it doesn't covert inputs to fp16 in mixed precision, it sends inputs to Linear in fp32 and converts them on the fly to fp16, returning fp16 result.

That's said: over at Deepspeed we are discussing automatically converting fp32 to fp16 under fp16 enabled and having a special flag to prevent that from happening should someone not want that behavior.

Comment (sgugger) [reply src/transformers/trainer.py:1711]: I am just talking from a user's perspective. Even if it's inside the CUDA kernel (though I think it's before that ;-) ), AMP does convert the input to FP16 at some point to do the operation, my main point is: the user does not see this and does not have to deal with it. :-)

Comment (stas00) [reply src/transformers/trainer.py:1711]: I wrote my comment after I downloaded their example and stepped through with debugger. The input to Linear is fp32 under amp and the conversion happens on C-level.

But I hear you, in the general case 99% of the time it should be safe for deepspeed to do the conversion automatically to match amp/apex.

Thank you for making the suggestion, @sgugger. I made a proposal here: https://github.com/microsoft/DeepSpeed/issues/1144

I was thinking more about the future where we have many more dtypes than just fp16 and fp32.

**Review by patrickvonplaten — APPROVED** [review]
That's awesome! Thanks so much for working on it :-)

**Review by sgugger — APPROVED** [review]
Just two nits, but looks great!

File: src/transformers/deepspeed.py [hunk 2]
```diff
@@ -70,46 +74,90 @@ def __init__(self, config_file_or_dict):
...
+        """
+        Returns :obj:`True`/:obj:`False` only if the value is set, always :obj:`False` otherwise. So use this method to
+        ask the very specific question of whether the value is set to :obj:`False` (and it's not set to :obj:`True` or
+        isn't set).
+        """
+        value = self.get_value(ds_key_long)
+        if value is None:
             return False
-        return bool(config.get(key))
+        return not bool(value)
```
[the remark this answers is not in the document]
Comment (stas00) [reply src/transformers/deepspeed.py:135]: neat, thank you!

**Comment by stas00** [comment]
I've just noticed that `Wav2Vec2Encoder` also uses `LayerDrop`
https://github.com/huggingface/transformers/blob/0e82f0cbc28b41b3d87a5e4069dc0e20bacc2494/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L555

So it'd break under multi-gpus - how do I activate this path so that it can be triggered and tested?

I did the synchronization for `Wav2Vec2EncoderStableLayerNorm` which was getting used by any current wav2vec2 script I tried.
